production/src/synch/course.rs

```diff
@@ -0,0 +1,82 @@
+use log::error;
+use super::sample_db_models;
+use stores::prelude::PgPool;
+use openapi::models;
+use super::util::extract_value_or_get_null_string;
+use std::fs::{OpenOptions, File, create_dir_all};
+use std::io::{BufWriter, Write};
+use std::path::Path;
+
+pub async fn get_courses_from_table(pool: &PgPool) -> Vec<sample_db_models::Course> {
+    let courses_from_table: Result<Vec<sample_db_models::Course>, sqlx::Error> = 
+        sqlx::query_as::<_, sample_db_models::Course>("SELECT * FROM courses")
+        .fetch_all(pool)
+        .await;
+
+    match courses_from_table {
+        Ok(courses) => courses,
+        Err(err) => {
+            error!("{:?}", err);
+            Vec::<sample_db_models::Course>::new()
+        }
+    }
+}
+
+/// Maps list of courses to an insert query for the courses table.
+pub fn map_courses_to_insert_query(courses: &Vec<models::Course>, file_path: &str) {
+    let path = Path::new(file_path);
+    if let Some(parent) = path.parent() {
+        create_dir_all(parent).expect("Failed to create intermediate directories");
+    }
+
+    let file: File = OpenOptions::new()
+        .write(true)
+        .create(true)
+        .open(file_path).expect("Could not open file");
+
+    let mut file: BufWriter<File> = BufWriter::new(file);
+
+    let start_of_query: &str = "INSERT INTO courses (catalog_number, subject_code, external_id \
+        , academic_level, title, description, requirements, enroll_consent, drop_consent, \
+        prerequisites_id)\nVALUES\n";
+    let fail_msg: &str = "Could not write insert query for courses table.";
+    file.write_all(start_of_query.as_bytes()).expect(fail_msg);
+
+    for (idx, course) in courses.iter().enumerate() {
+        let mut val: String = map_course_to_query(course);
+        // For the last tuple, we don't want a comma at the end.
+        if idx == courses.len() - 1 { 
+            val.pop(); // remove newline.
+            val.pop(); // remove comma.
+            val.push('\n'); // insert newline again.
+        }
+        file.write_all(val.as_bytes()).expect(fail_msg);
+    }
+
+    // For now, just set it as do nothing.
+    let end_of_query: &str = "ON CONFLICT DO NOTHING;";
```

Comment:
Might want to update set with the newest course entry if there is a conflict? In case a course does get updated?

Reply:
Yeah, I wanted to do this, but turns out it causes an error for the bulk insert because for the "full" version, we get **duplicate** entries with the SAME course code AND catalog number, but different external course id. So this basically causes the error: 

`ERROR:  ON CONFLICT DO UPDATE command cannot affect row a second time
HINT:  Ensure that no rows proposed for insertion within the same command have duplicate constrained values.`

So this would have required additional logic for uniqueness of courses when writing the insert query. And I didn't want to deal w/ that, while I was still working on the core logic for the synchronizer. But I can probably update this now.